javascript/ql/src/semmle/javascript/dependencies/SemVer.qll

```diff
@@ -0,0 +1,103 @@
+
+/**
+ * Provides classes for working SemVer (Semantic Versioning).
+ */
+
+import semmle.javascript.dependencies.Dependencies
+
+/**
+ * A SemVer-formatted version string in a dependency.
+ */
+class DependencySemVer extends string {
+  Dependency dep;
+
+  int major;
+
+  int minor;
+
+  int patch;
+
+  string ext;
```

Comment:
I think this could be simplified substantially by mapping semver strings to a "padded" semver string whose lexicographical ordering is consistent with semver ordering. For example, `4.17.11` is mapped to `004.017.011` after padding.

I used this dirty version for a bit, which was nice since you could write things like `version < semver("4.17.11")`
```
  bindingset[str]
  string leftPad(string str) { result = ("000" + str).suffix(str.length()) }

  bindingset[semver]
  string semver(string semver) {
    result = leftPad(semver.splitAt(".", 0)) + "." + leftPad(semver.splitAt(".", 1)) + "." +
        leftPad(semver.splitAt(".", 2))
  }
```

This version doesn't handle every case, but the idea could be used to avoid the scary bindingsets and disjunctions in this class.

WDYT?